client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx

```diff
@@ -82,6 +88,20 @@ function PaymentMethodAdd(): ReactElement {
 		[ reduxDispatch ]
 	);
 
+	const successCallback = useCallback( () => {
+		if ( getQueryArg( window.location.href, 'return' ) ) {
+			reduxDispatch(
+				fetchStoredCards( {
+					startingAfter: '',
+					endingBefore: '',
+				} )
+			);
+			setSuccessRedirect( () => true );
+		} else {
+			page( '/partner-portal/payment-methods/' );
+		}
+	}, [ reduxDispatch, page, setSuccessRedirect, window, getQueryArg ] );
```

Comment:
The fix doesn't seem to work for me. I'm having trouble following the logic here but it seems like we're relying on a race condition where we fetch the payment methods and redirect and then hope we get the response before the user takes an action.

IMO we need to make this more robust by doing something like this:
- Update `useReturnUrl()` so it does not attempt to redirect if there's no `return` argument even if the passed parameter is true.
- Use `hasValidPaymentMethod()` here and pass it's return value to useReturnUrl() so that if there is a return URL and there is a valid payment method useReturnUrl will trigger and redirect. This will make `successRedirect` unnecessary.
- Since we now use `hasValidPaymentMethod()` our component will re-render when that value changes thus triggering the redirect.

<details>
  <summary>Here's a quick diff - I have just barely tested this so don't rely on it :D</summary>
  
```diff
diff --git a/client/jetpack-cloud/sections/partner-portal/hooks.ts b/client/jetpack-cloud/sections/partner-portal/hooks.ts
index 7b55056c93..7cdd0f9574 100644
--- a/client/jetpack-cloud/sections/partner-portal/hooks.ts
+++ b/client/jetpack-cloud/sections/partner-portal/hooks.ts
@@ -14,6 +14,11 @@ export function useReturnUrl( redirect: boolean ): void {
 	useEffect( () => {
 		if ( redirect ) {
 			const returnQuery = getQueryArg( window.location.href, 'return' ) as string;
+
+			if ( ! returnQuery ) {
+				return;
+			}
+
 			const returnUrl = ensurePartnerPortalReturnUrl( returnQuery );
 
 			page.redirect( returnUrl );
diff --git a/client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx b/client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx
index e509043159..68545cf841 100644
--- a/client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx
+++ b/client/jetpack-cloud/sections/partner-portal/primary/payment-method-add/index.tsx
@@ -28,6 +28,7 @@ import SidebarNavigation from 'calypso/jetpack-cloud/sections/partner-portal/sid
 import { recordTracksEvent } from 'calypso/state/analytics/actions';
 import { getCurrentUserLocale } from 'calypso/state/current-user/selectors';
 import { errorNotice, removeNotice, successNotice } from 'calypso/state/notices/actions';
+import { hasValidPaymentMethod } from 'calypso/state/partner-portal/partner/selectors';
 import { fetchStoredCards } from 'calypso/state/partner-portal/stored-cards/actions';
 
 import './style.scss';
@@ -35,7 +36,7 @@ import './style.scss';
 function PaymentMethodAdd(): ReactElement {
 	const translate = useTranslate();
 	const reduxDispatch = useDispatch();
-	const [ successRedirect, setSuccessRedirect ] = useState( false );
+	const hasPaymentMethod = useSelector( hasValidPaymentMethod );
 	const { isStripeLoading, stripeLoadingError, stripeConfiguration, stripe } = useStripe();
 	const {
 		reload: reloadSetupIntentId,
@@ -56,7 +57,7 @@ function PaymentMethodAdd(): ReactElement {
 		select( 'credit-card' ).useAsPrimaryPaymentMethod()
 	);
 
-	useReturnUrl( successRedirect );
+	useReturnUrl( hasPaymentMethod );
 
 	const onGoToPaymentMethods = () => {
 		reduxDispatch(
@@ -96,11 +97,10 @@ function PaymentMethodAdd(): ReactElement {
 					endingBefore: '',
 				} )
 			);
-			setSuccessRedirect( () => true );
 		} else {
 			page( '/partner-portal/payment-methods/' );
 		}
-	}, [ reduxDispatch, page, setSuccessRedirect, window, getQueryArg ] );
+	}, [ reduxDispatch, page, window, getQueryArg ] );
 
 	useEffect( () => {
 		if ( stripeLoadingError ) {
```
</details>

Reply:
This is a great idea! 
I haven't counted for the race condition, completely missed that point. I applied your suggestion and it seems to work wonders, now when successfully adding a credit card it properly waits for the `hasValidPaymentMethod` flag to be true before redirecting.